File: packages/sdk/src/otel.ts
/**
 * Convenience wiring for AcruxCore's OTLP trace ingestion endpoint.
 *
 * AcruxCore accepts traces from any OpenTelemetry (OTel) source at
 * `POST /api/v1/traces/otlp` — the OpenAI Agents SDK, LangChain, or a hand-rolled OTel
 * pipeline all work with nothing but a `TracerProvider`, a `BatchSpanProcessor`, and an
 * OTLP exporter pointed at that endpoint. This module is optional sugar over that same
 * wiring: it does not change what the endpoint accepts, and a hand-written pipeline
 * works identically.
 *
 * The OTel packages this module imports are optional peer dependencies of
 * `@acruxcoreai/sdk` — install them alongside the SDK to use `register()`:
 * `npm install @opentelemetry/api @opentelemetry/sdk-trace-node @opentelemetry/sdk-trace-base
 * @opentelemetry/exporter-trace-otlp-http @opentelemetry/resources @opentelemetry/semantic-conventions`.
 * Framework auto-instrumentation via `instrument: [...]` additionally requires that
 * framework's own package and its `@arizeai/openinference-instrumentation-*` package;
 * {@link register} raises a clear error naming the missing package rather than
 * installing it for you.
 */

import { acruxcoreError } from './error';

/** One entry in {@link SUPPORTED_FRAMEWORKS}'s backing registry. */
interface InstrumentorEntry {
  /** The framework's own npm package — dynamically imported and handed to `manuallyInstrument()`. */
  frameworkPackage: string;
  /** The `@arizeai/openinference-instrumentation-*` package providing the instrumentation class. */
  instrumentationPackage: string;
  /** Name of the exported `InstrumentationBase` subclass within `instrumentationPackage`. */
  className: string;
}

const INSTRUMENTOR_REGISTRY = {
  openai: {
    frameworkPackage: 'openai',
    instrumentationPackage: '@arizeai/openinference-instrumentation-openai',
    className: 'OpenAIInstrumentation',
  },
  openai_agents: {
    frameworkPackage: '@openai/agents',
    instrumentationPackage: '@arizeai/openinference-instrumentation-openai-agents',
    className: 'OpenAIAgentsInstrumentation',
  },
} as const satisfies Record<string, InstrumentorEntry>;

/**
 * Framework names accepted by `register({ instrument: [...] })`, sorted for display.
 *
 * Shorter than the Python SDK's list by design: LangChain.js's OpenInference
 * instrumentor patches `@langchain/core/callbacks/manager` rather than the top-level
 * package, and LlamaIndex.TS's `@arizeai/openinference-instrumentation-llama-index` is
 * an empty placeholder as of `0.0.12` — neither has a stable wiring to hardcode yet.
 * Both frameworks still work with {@link register}'s bare pipeline plus their own
 * hand-wired instrumentor, the same as any framework outside this list.
 */
export const SUPPORTED_FRAMEWORKS = Object.keys(INSTRUMENTOR_REGISTRY).sort() as SupportedFramework[];

/** A framework name accepted by `register({ instrument: [...] })`. */
export type SupportedFramework = keyof typeof INSTRUMENTOR_REGISTRY;

/** Options for {@link register}. */
export interface RegisterOptions {
  /** AcruxCore API key. Falls back to `ACRUXCORE_API_KEY`. */
  apiKey?: string;
  /** AcruxCore API base URL, e.g. `https://api.acruxcore.com/api/v1`. Falls back to `ACRUXCORE_BASE_URL`. */
  baseUrl?: string;
  /** Reported as the OTel `service.name` resource attribute. Defaults to `acruxcore-instrumented-app`. */
  serviceName?: string;
  /**
   * Framework names to auto-instrument against the returned provider, e.g.
   * `['openai_agents']`. See {@link SUPPORTED_FRAMEWORKS} for the full list. Each
   * name's framework package and OpenInference instrumentation package must already
   * be installed; this only instruments them, it never installs anything.
   */
  instrument?: SupportedFramework[];
  /**
   * Also install the provider process-wide via `NodeTracerProvider#register()`
   * (default `true`) — not just `trace.setGlobalTracerProvider`, but also an
   * `AsyncLocalStorageContextManager` and a W3C propagator. Without that context
   * manager, OTel context (parent spans, `context.with()` session grouping) does
   * not survive an `await`, so turning this off requires wiring an equivalent
   * context manager yourself before any instrumented async code runs.
   */
  setGlobal?: boolean;
}

/**
 * Builds an OTel `NodeTracerProvider` that exports straight to AcruxCore.
 *
 * Collapses the `NodeTracerProvider` + `BatchSpanProcessor` + `OTLPTraceExporter`
 * wiring every OTLP tracing tutorial hand-writes into one call, and optionally
 * instruments a named framework's own package against the resulting provider.
 *
 * OTel's own instrumentation classes patch a framework via `require()` hooks, which
 * never fire for an ESM `import` — so this dynamically imports each requested
 * framework's package itself and passes it to the instrumentor's
 * `manuallyInstrument()`, which works in both module systems. Because Node resolves a
 * given package to one cached module instance per process, the object patched here is
 * the same one any later `import` in the caller's own code sees.
 *
 * @param options - See {@link RegisterOptions}.
 * @returns The configured `NodeTracerProvider` — pass it to instrumentors you wire
 *   yourself, or to `@arizeai/openinference-core`'s `context.with(setSession(...))`.
 * @throws {acruxcoreError} `OTEL_NOT_AVAILABLE` if the OTel peer dependencies aren't
 *   installed; `MISSING_API_KEY` / `MISSING_BASE_URL` if required config is absent;
 *   `UNKNOWN_INSTRUMENTOR` for a name outside {@link SUPPORTED_FRAMEWORKS};
 *   `INSTRUMENTOR_NOT_INSTALLED` if that framework's package or OpenInference package
 *   isn't installed.
 */
export async function register(options: RegisterOptions = {}) {
  let sdkTraceNode: typeof import('@opentelemetry/sdk-trace-node');
  let sdkTraceBase: typeof import('@opentelemetry/sdk-trace-base');
  let exporterOtlpHttp: typeof import('@opentelemetry/exporter-trace-otlp-http');
  let resources: typeof import('@opentelemetry/resources');
  let semconv: typeof import('@opentelemetry/semantic-conventions');
  try {
    // @opentelemetry/api itself is not imported directly here — sdk-trace-node
    // requires it internally, so a missing install already fails this Promise.all.
    [sdkTraceNode, sdkTraceBase, exporterOtlpHttp, resources, semconv] = await Promise.all([
      import('@opentelemetry/sdk-trace-node'),
      import('@opentelemetry/sdk-trace-base'),
      import('@opentelemetry/exporter-trace-otlp-http'),
      import('@opentelemetry/resources'),
      import('@opentelemetry/semantic-conventions'),
    ]);
  } catch (exc) {
    throw new acruxcoreError(
      'acruxcore/otel register() needs the OTel peer dependencies: npm install ' +
        '@opentelemetry/api @opentelemetry/sdk-trace-node @opentelemetry/sdk-trace-base ' +
        '@opentelemetry/exporter-trace-otlp-http @opentelemetry/resources @opentelemetry/semantic-conventions',
      'OTEL_NOT_AVAILABLE',
    );
  }

  const resolvedKey = options.apiKey ?? process.env.ACRUXCORE_API_KEY;
  if (!resolvedKey) {
    throw new acruxcoreError(
      'acruxcore/otel register(): apiKey is required. Pass it or set ACRUXCORE_API_KEY.',
      'MISSING_API_KEY',
    );
  }

  const resolvedBase = options.baseUrl ?? process.env.ACRUXCORE_BASE_URL;
  if (!resolvedBase) {
    throw new acruxcoreError(
      'acruxcore/otel register(): baseUrl is required. Pass it or set ACRUXCORE_BASE_URL.',
      'MISSING_BASE_URL',
    );
  }

  const endpoint = `${resolvedBase.replace(/\/+$/, '')}/traces/otlp`;
  const exporter = new exporterOtlpHttp.OTLPTraceExporter({
    url: endpoint,
    headers: { Authorization: `Bearer ${resolvedKey}` },
  });
  const provider = new sdkTraceNode.NodeTracerProvider({
    resource: resources.resourceFromAttributes({
      [semconv.ATTR_SERVICE_NAME]: options.serviceName ?? 'acruxcore-instrumented-app',
    }),
    spanProcessors: [new sdkTraceBase.BatchSpanProcessor(exporter)],
  });

  if (options.setGlobal ?? true) {
    // NodeTracerProvider#register() does more than trace.setGlobalTracerProvider():
    // it also installs an AsyncLocalStorageContextManager and a W3C propagator.
    // Without that context manager, OTel context (parent spans, `context.with()`
    // session grouping) does not survive an `await` — it would silently stop
    // propagating past the first async boundary in any real agent framework.
    provider.register();
  }

  for (const name of options.instrument ?? []) {
    await instrumentFramework(name, provider);
  }

  return provider;
}

/** Looks up and instruments one framework by name against the given provider. */
async function instrumentFramework(name: SupportedFramework, provider: unknown): Promise<void> {
  const entry = (INSTRUMENTOR_REGISTRY as Record<string, InstrumentorEntry>)[name];
  if (!entry) {
    throw new acruxcoreError(
      `acruxcore/otel: unknown framework ${JSON.stringify(name)}. Supported: ${SUPPORTED_FRAMEWORKS.join(', ')}.`,
      'UNKNOWN_INSTRUMENTOR',
    );
  }

  let instrumentationModule: Record<string, unknown>;
  try {
    instrumentationModule = (await import(entry.instrumentationPackage)) as Record<string, unknown>;
  } catch (exc) {
    throw new acruxcoreError(
      `acruxcore/otel: instrument=${JSON.stringify(name)} needs "${entry.instrumentationPackage}". ` +
        `Install it with: npm install ${entry.instrumentationPackage}`,
      'INSTRUMENTOR_NOT_INSTALLED',
    );
  }

  let frameworkModule: unknown;
  try {
    frameworkModule = await import(entry.frameworkPackage);
  } catch (exc) {
    throw new acruxcoreError(
      `acruxcore/otel: instrument=${JSON.stringify(name)} needs "${entry.frameworkPackage}". ` +
        `Install it with: npm install ${entry.frameworkPackage}`,
      'INSTRUMENTOR_NOT_INSTALLED',
    );
  }

  const InstrumentationClass = instrumentationModule[entry.className] as new (opts: {
    tracerProvider: unknown;
  }) => { manuallyInstrument(mod: unknown): void };
  const instrumentation = new InstrumentationClass({ tracerProvider: provider });
  instrumentation.manuallyInstrument(frameworkModule);
}

File: packages/sdk-python/src/acruxcore/otel.py
"""Convenience wiring for AcruxCore's OTLP trace ingestion endpoint.

AcruxCore accepts traces from any OpenTelemetry (OTel) source at
``POST /api/v1/traces/otlp`` — CrewAI, the OpenAI Agents SDK, LangChain, LlamaIndex,
or a hand-rolled OTel pipeline all work with nothing but a ``TracerProvider``, a
``BatchSpanProcessor``, and an OTLP exporter pointed at that endpoint. This module
is optional sugar over that same wiring: it does not change what the endpoint
accepts, and a hand-written pipeline (as shown in the OTLP tracing tutorials) works
identically.

Requires the ``otel`` extra — ``pip install acruxcore[otel]`` — since
``opentelemetry-sdk`` is not a runtime dependency of the base package. Framework
auto-instrumentation via ``instrument=[...]`` additionally requires that
framework's own ``openinference-instrumentation-*`` package; :func:`register`
raises a clear error naming the missing package rather than installing it for you.
"""

from __future__ import annotations

import importlib
import os
from typing import TYPE_CHECKING, Dict, Optional, Sequence, Tuple

from .errors import (
    INSTRUMENTOR_NOT_INSTALLED,
    MISSING_API_KEY,
    MISSING_BASE_URL,
    OTEL_NOT_AVAILABLE,
    UNKNOWN_INSTRUMENTOR,
    AcruxCoreError,
)

if TYPE_CHECKING:
    from opentelemetry.sdk.trace import TracerProvider

#: module path, OpenInference instrumentor class name, pip package to install.
_InstrumentorEntry = Tuple[str, str, str]

_INSTRUMENTOR_REGISTRY: Dict[str, _InstrumentorEntry] = {
    "openai": (
        "openinference.instrumentation.openai",
        "OpenAIInstrumentor",
        "openinference-instrumentation-openai",
    ),
    "openai_agents": (
        "openinference.instrumentation.openai_agents",
        "OpenAIAgentsInstrumentor",
        "openinference-instrumentation-openai-agents",
    ),
    "crewai": (
        "openinference.instrumentation.crewai",
        "CrewAIInstrumentor",
        "openinference-instrumentation-crewai",
    ),
    "langchain": (
        "openinference.instrumentation.langchain",
        "LangChainInstrumentor",
        "openinference-instrumentation-langchain",
    ),
    "llama_index": (
        "openinference.instrumentation.llama_index",
        "LlamaIndexInstrumentor",
        "openinference-instrumentation-llama-index",
    ),
}

#: Framework names accepted by ``register(instrument=[...])``, sorted for display.
SUPPORTED_FRAMEWORKS: Tuple[str, ...] = tuple(sorted(_INSTRUMENTOR_REGISTRY))


def register(
    api_key: Optional[str] = None,
    base_url: Optional[str] = None,
    *,
    service_name: str = "acruxcore-instrumented-app",
    instrument: Optional[Sequence[str]] = None,
    set_global: bool = True,
) -> "TracerProvider":
    """Build an OTel ``TracerProvider`` that exports straight to AcruxCore.

    Collapses the ``TracerProvider`` + ``BatchSpanProcessor`` + ``OTLPSpanExporter``
    wiring every OTLP tracing tutorial hand-writes into one call, and optionally
    turns on a named framework's own OpenInference instrumentor against the
    resulting provider.

    :param api_key: AcruxCore API key. Fallback: ``ACRUXCORE_API_KEY``.
    :param base_url: AcruxCore API base URL, e.g.
        ``https://api.acruxcore.com/api/v1``. Fallback: ``ACRUXCORE_BASE_URL``.
    :param service_name: Reported as the OTel ``service.name`` resource attribute.
    :param instrument: Framework names to auto-instrument against the returned
        provider, e.g. ``["crewai", "openai"]`` — CrewAI calls the plain OpenAI
        SDK by default, so both are needed together for cost/token data. See
        :data:`SUPPORTED_FRAMEWORKS` for the full list. Each name's
        ``openinference-instrumentation-*`` package must already be installed;
        this only calls its ``.instrument(tracer_provider=...)``, it never
        installs anything.
    :param set_global: Also install the provider as the process-wide default via
        ``opentelemetry.trace.set_tracer_provider`` (default ``True``), so an
        instrumentor or manual ``trace.get_tracer(...)`` call that doesn't take
        an explicit provider still exports to AcruxCore.
    :returns: The configured ``TracerProvider`` — pass it to instrumentors you
        wire yourself, or to ``openinference.instrumentation.using_session``.
    :raises AcruxCoreError: ``OTEL_NOT_AVAILABLE`` if the ``otel`` extra isn't
        installed; ``MISSING_API_KEY`` / ``MISSING_BASE_URL`` if required config
        is absent; ``UNKNOWN_INSTRUMENTOR`` for a name outside
        :data:`SUPPORTED_FRAMEWORKS`; ``INSTRUMENTOR_NOT_INSTALLED`` if that
        framework's OpenInference package isn't installed.
    """
    try:
        from opentelemetry import trace
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
            OTLPSpanExporter,
        )
        from opentelemetry.sdk.resources import SERVICE_NAME, Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
    except ImportError as exc:
        raise AcruxCoreError(
            "acruxcore.otel.register() needs the 'otel' extra: "
            "pip install acruxcore[otel]",
            OTEL_NOT_AVAILABLE,
        ) from exc

    resolved_key = api_key or os.environ.get("ACRUXCORE_API_KEY")
    if not resolved_key:
        raise AcruxCoreError(
            "acruxcore.otel.register(): api_key is required. Pass it or set "
            "ACRUXCORE_API_KEY.",
            MISSING_API_KEY,
        )

    resolved_base = base_url or os.environ.get("ACRUXCORE_BASE_URL")
    if not resolved_base:
        raise AcruxCoreError(
            "acruxcore.otel.register(): base_url is required. Pass it or set "
            "ACRUXCORE_BASE_URL.",
            MISSING_BASE_URL,
        )

    endpoint = f"{resolved_base.rstrip('/')}/traces/otlp"
    provider = TracerProvider(resource=Resource.create({SERVICE_NAME: service_name}))
    exporter = OTLPSpanExporter(
        endpoint=endpoint,
        headers={"Authorization": f"Bearer {resolved_key}"},
    )
    provider.add_span_processor(BatchSpanProcessor(exporter))

    if set_global:
        trace.set_tracer_provider(provider)

    for name in instrument or ():
        _instrument_framework(name, provider)

    return provider


def _instrument_framework(name: str, provider: "TracerProvider") -> None:
    """Look up and call one framework's OpenInference instrumentor by name."""
    entry = _INSTRUMENTOR_REGISTRY.get(name)
    if entry is None:
        raise AcruxCoreError(
            f"acruxcore.otel: unknown framework {name!r}. Supported: "
            f"{', '.join(SUPPORTED_FRAMEWORKS)}.",
            UNKNOWN_INSTRUMENTOR,
        )
    module_path, class_name, pip_package = entry
    try:
        module = importlib.import_module(module_path)
    except ImportError as exc:
        raise AcruxCoreError(
            f"acruxcore.otel: instrument={name!r} needs {pip_package!r}. Install "
            f"it with: pip install {pip_package}",
            INSTRUMENTOR_NOT_INSTALLED,
        ) from exc
    instrumentor_cls = getattr(module, class_name)
    instrumentor_cls().instrument(tracer_provider=provider)
